cmd/rook/util/secret.go

```diff
@@ -0,0 +1,110 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"os/signal"
+
+	"github.com/pkg/errors"
+	"github.com/rook/rook/cmd/rook/rook"
+	"github.com/rook/rook/pkg/daemon/ceph/client"
+	"github.com/rook/rook/pkg/daemon/ceph/osd/kms"
+	operator "github.com/rook/rook/pkg/operator/ceph"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// SecretCmd defines a top-level utility command which interacts with secret
+var SecretCmd = &cobra.Command{
+	Use:   "secret",
+	Short: "Secret interacts with a given Key Management System and perform actions.",
+	Long: `The secret sub-command helps interacting with Key Management System.
+	It can perform various actions such as retrieving the content of a Key Encryption Key.`,
+}
+
+func init() {
+	SecretCmd.AddCommand(
+		cliGetSecret(),
+	)
+}
+
+func startSecret() *kms.Config {
+	// Initialize the context
+	ctx, cancel := signal.NotifyContext(context.Background(), operator.ShutdownSignals...)
+	defer cancel()
+
+	namespace := os.Getenv(k8sutil.PodNamespaceEnvVar)
+	if namespace == "" {
+		rook.TerminateFatal(errors.New("failed to find pod namespace"))
+	}
+
+	name := os.Getenv("ROOK_CLUSTER_NAME")
+	if name == "" {
+		rook.TerminateFatal(errors.New("failed to find cluster's name"))
+	}
+
+	clusterInfo := client.NewClusterInfo(namespace, name)
+	clusterInfo.Context = ctx
+	context := rook.NewContext()
+
+	// Fetch the CephCluster for the KMS details
+	cephCluster, err := context.RookClientset.CephV1().CephClusters(namespace).Get(ctx, namespace, metav1.GetOptions{})
+	if err != nil {
+		rook.TerminateFatal(errors.Wrapf(err, "failed to get ceph cluster in namespace %q", namespace))
+	}
+
+	// Validate connection details
+	err = kms.ValidateConnectionDetails(context, &cephCluster.Spec.Security, namespace)
+	if err != nil {
+		rook.TerminateFatal(errors.Wrap(err, "failed to validate kms connection details"))
+	}
+
+	return kms.NewConfig(context, &cephCluster.Spec, clusterInfo)
+}
+
+// cliGetSecret is the Cobra CLI call
+func cliGetSecret() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "get [name] [path]",
+		Short: "Fetch a secret",
+		Args:  cobra.ExactArgs(2),
+		Run:   getSecret,
+	}
+	return cmd
+}
+
+func getSecret(cmd *cobra.Command, args []string) {
+	secretName := args[0]
+	secretPath := args[1]
+	config := startSecret()
+
+	// Fetch the secret
+	s, err := config.GetSecret(secretName)
```

Comment:
Why do we have to get the secret like this? Why can't the secret be mounted as a volume to the osd pod, just like we use other secrets and configmaps?

Reply:
Because Vault Secrets are not Kubernetes objects. "secret" is the generic interface name but secret can be K8s Secret, or Vault, any KMS.

Reply:
Ah of course

Reply:
I think maybe we would benefit from calling this `GetKMSSecret` so we know it is not  a Kubernetes Secret. I think since we work with K8s often, there will be a lot of misunderstanding/conflating like this in the future as well.